### PR TITLE
Flutter web support structure init + implement amplitude init and logEvent calls

### PR DESCRIPTION
**Pull request by yuhao900914**

### CODE REVIEW

File: lib/amplitude_web.dart [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          return true;
+        }
+      case "logEvent":
+        {
+          var eventType = args['eventType'];
+          amplitude.logEvent(eventType);
+          return true;
+        }
+      default:
+      //(FlutterMethodNotImplemented)
```
Comment (haoliu-amp) [inline lib/amplitude_web.dart]: throw an exception instead of a comment?

**Review by haoliu-amp — CHANGES_REQUESTED** [review]
The title of the PR makes me feel you finished the bet. :) Could be more descriptive.

File: example/web/index.html [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+  <meta charset="UTF-8">
+  <meta content="IE=Edge" http-equiv="X-UA-Compatible">
+  <meta name="description" content="A new Flutter project.">
+
+  <!-- iOS meta tags & icons -->
+  <meta name="apple-mobile-web-app-capable" content="yes">
+  <meta name="apple-mobile-web-app-status-bar-style" content="black">
+  <meta name="apple-mobile-web-app-title" content="example">
+  <link rel="apple-touch-icon" href="icons/Icon-192.png">
```
Comment (qingzhuozhen) [inline example/web/index.html]: Can't find a way to comment on icon files, so just put here, do we need these icons or images?

File: example/web/index.html [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+  <title>example</title>
+  <link rel="manifest" href="manifest.json">
+</head>
+<body>
+
+  <script type="text/javascript" defer>
+    console.log("in lib script snippeit");
```
Comment (haoliu-amp) [inline example/web/index.html]: typo.

File: lib/amplitude_web.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        {
+          var apiKey = args['apiKey'];
+          amplitude.init(apiKey, null, {});
+          return true;
+        }
+      case "logEvent":
+        {
+          var eventType = args['eventType'];
+          amplitude.logEvent(eventType);
+          return true;
```
Comment (qingzhuozhen) [inline lib/amplitude_web.dart]: For other arguments, is the plan to complete them in PR with other setter methods?

File: example/web/index.html (hunk 3 above)
Comment (haoliu-amp) [reply example/web/index.html]: Do you need this log also

File: example/web/index.html [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+      (function (e, t) {
+        var n = e.amplitude || { _q: [], _iq: {} }; var r = t.createElement("script")
+          ; r.type = "text/javascript"
+          ; r.integrity = "sha384-9jIJQlEp6cEk6foaEO/N5yzaJmQbTtugbpg9l8P/javk55WlPh7fJGR1vHhyDjON"
+          ; r.crossOrigin = "anonymous"; r.async = true
+          ; r.src = "https://cdn.amplitude.com/libs/amplitude-8.4.0-min.gz.js"
+          ; r.onload = function () {
+            if (!e.amplitude.runQueuedFunctions) {
+        console.log("[Amplitude] Error: could not load SDK")
```
Comment (haoliu-amp) [inline example/web/index.html:41]: Should it be console.error?

File: example/web/index.html [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+              loadMainDartJs();
+            }
+          });
+
+        // If service worker doesn't succeed in a reasonable amount of time,
+        // fallback to plaint <script> tag.
+        setTimeout(() => {
+          if (!scriptLoaded) {
+            console.warn(
+              'Failed to load app from service worker. Falling back to plain <script> tag.',
```
Comment (haoliu-amp) [inline example/web/index.html]: Prefix with [Amplitude]

File: example/web/index.html (hunk 3 above)
Comment (haoliu-amp) [reply example/web/index.html]: If you do, prefix by [Amplitude]

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

File: example/web/index.html (hunk 3 above)
Comment (yuhao900914) [reply example/web/index.html]: We don't need this actually, That's for my testing. Maybe I will have a code clean-up after all is finished.

File: lib/amplitude_web.dart [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    var args = jsonDecode(call.arguments.toString());
+    var instanceName = args['instanceName'];
+    Amplitude amplitude = Amplitude.getInstance(instanceName);
+
+    switch (call.method) {
+      // Init
+      case "init":
+        {
+          var apiKey = args['apiKey'];
+          amplitude.init(apiKey, null, {});
```
Comment (qingzhuozhen) [inline lib/amplitude_web.dart]: this method does have the user id right? Do we want to cover it as well or do it together in implantation PR?

File: lib/amplitude_web.dart [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+import 'package:js/js.dart';
+import 'dart:convert';
+
+import 'package:flutter/services.dart';
+import 'package:flutter_web_plugins/flutter_web_plugins.dart';
+import 'dart:async';
+
+@JS('amplitude')
+class Amplitude {
```
Comment (qingzhuozhen) [inline lib/amplitude_web.dart]: It is ok to keep this here for now. Just wonder if in future we have a long list of methods, should we move this JS interop into another file, then imported into this file?

File: example/web/index.html (hunk 5 above)
Comment (yuhao900914) [reply example/web/index.html:41]: I copied this from our developer center, do I also need to change it there?

File: example/web/index.html [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        // Otherwise, the browser will load the script multiple times,
+        // potentially different versions.
+        var serviceWorkerUrl = 'flutter_service_worker.js?v=' + serviceWorkerVersion;
+        navigator.serviceWorker.register(serviceWorkerUrl)
+          .then((reg) => {
+            function waitForActivation(serviceWorker) {
+              serviceWorker.addEventListener('statechange', () => {
+                if (serviceWorker.state == 'activated') {
+                  console.log('Installed new service worker.');
+                  loadMainDartJs();
```
Comment (qingzhuozhen) [inline example/web/index.html:106]: Just curious does this work in service worker? I remember we have our customers asked about JS SDK support in service worker.

File: lib/amplitude_web.dart (hunk 8 above)
Comment (yuhao900914) [reply lib/amplitude_web.dart]: I think we can, will re-organize this.

File: example/web/index.html (hunk 2 above)
Comment (yuhao900914) [reply example/web/index.html]: Actually, that's all auto-generated.  I Will try if we can clean this up.

File: example/web/index.html (hunk 5 above)
Comment (haoliu-amp) [reply example/web/index.html:41]: If it's from the snippet, that's fine.

File: lib/amplitude_web.dart [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  /// https://flutter.dev/go/federated-plugins
+  Future<dynamic> handleMethodCall(MethodCall call) async {
+    var args = jsonDecode(call.arguments.toString());
+    var instanceName = args['instanceName'];
+    Amplitude amplitude = Amplitude.getInstance(instanceName);
+    switch (call.method) {
+      case "init":
+        {
+          String apiKey = args['apiKey'];
+          return amplitude.init(apiKey, args['userId'] ?? null);
```
Comment (qingzhuozhen) [inline lib/amplitude_web.dart]: nit: this default behavior could be moved out of directly calling

File: lib/amplitude_web.dart [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        }
+      case "logEvent":
+        {
+          String eventType = args['eventType'];
+          return amplitude.logEvent(eventType, args['eventProperties'] ?? null);
```
Comment (qingzhuozhen) [inline lib/amplitude_web.dart]: Does current example covers event properties as well? If not we can cover it in next PR.

**Review by qingzhuozhen — APPROVED** [review]

File: lib/amplitude_web.dart (hunk 10 above)
Comment (yuhao900914) [reply lib/amplitude_web.dart]: fixed

File: lib/amplitude_web.dart (hunk 11 above)
Comment (yuhao900914) [reply lib/amplitude_web.dart]: Yes, it's in the example app. The dart Map needs to convert to JSObject before passing into the javascript function. I just made new changes to fix the eventProperties can not be passed into JS.

Comment (yuhao900914) [reply lib/amplitude_web.dart]: Screen Shot 2021-11-09 at 3 54 58 PM

File: lib/amplitude_web.dart [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        {
+          String apiKey = args['apiKey'];
+          var userId = args['userId'] ?? null;
+          return amplitude.init(apiKey, userId);
+        }
+      case "logEvent":
+        {
+          String eventType = args['eventType'];
+          var eventProperties = (args['eventProperties'] != null)
+              ? mapToJSObj(args['eventProperties'])
```
Comment (qingzhuozhen) [inline lib/amplitude_web.dart:38]: Good found!

**Review by qingzhuozhen — APPROVED** [review]

**Pull request merged**